File: MyLeetCode/FB/Insert Interval.py
###
# Time Complexity: O(n)
# Space Complexity: O(n)
###
# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution(object):
    def insert(self, intervals, newInterval):
        """
        :type intervals: List[Interval]
        :type newInterval: Interval
        :rtype: List[Interval]
        """
        if not intervals:
            return [newInterval]
        stack = []
        for e in intervals:
            if e.end < newInterval.start:
                stack.append(e)
            elif newInterval.end < e.start:
                stack.append(newInterval)
                newInterval = e
            else:
                newInterval.start = min(newInterval.start, e.start)
                newInterval.end = max(newInterval.end, e.end)
        stack.append(newInterval)
        return stack
                
        
        

File: MyLeetCode/Leetcode/Insert Interval.cpp
#if 1
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
    vector<Interval> insert(vector<Interval>& intervals, Interval newInterval) {
        vector<Interval> res;
        int n = intervals.size();
        if (n == 0) {
            res.push_back(newInterval);
            return res;
        }
        for (int i = 0; i < n; i++)
        {
            if (intervals[i].end < newInterval.start)
                res.push_back(intervals[i]);
            else if (newInterval.end < intervals[i].start)
            {
                res.push_back(newInterval);
                newInterval = intervals[i];
            }
            //overlapped
            else
            {
                newInterval.start = min(newInterval.start, intervals[i].start);
                newInterval.end = max(newInterval.end, intervals[i].end);
            }
        }
        res.push_back(newInterval);
        return res;
    }
};
#else
//no extra space but time limit exceed. guess resulting from erase() and insert();
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
#define MIN(a, b) ((a < b) ? (a) : (b))
#define MAX(a, b) ((a > b) ? (a) : (b))
class Solution {
public:
    vector<Interval> insert(vector<Interval> &intervals, Interval newInterval) {
        vector<Interval>::iterator it = intervals.begin();

        while (it != intervals.end())
        {
            if (it->end < newInterval.start)
                it++;
            else if (newInterval.end < it->start)
            {
                intervals.insert(it, newInterval);
                return intervals;
            }
            else
            {
                newInterval.start = MIN(newInterval.start, it->start);
                newInterval.end   = MAX(newInterval.end, it->end);
                it = intervals.erase(it);
            }
        }
        intervals.insert(it, newInterval);
        return intervals;
    }
};